diff --git a/CMakeLists.txt b/CMakeLists.txt
index adde7f2b..fdf5cc4b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -69,7 +69,8 @@ include_directories(
 
 if(ENABLE_SPM)
     execute_process(
-        COMMAND ${Python_EXECUTABLE} -c "from shedskin import cmake; cmake.user_cache_dir()"
+        COMMAND ${Python_EXECUTABLE} -m shedskin.cmake
+        # COMMAND ${Python_EXECUTABLE} -c "from shedskin import cmake; cmake.user_cache_dir()"
         OUTPUT_VARIABLE SPM_DEPS
         COMMAND_ERROR_IS_FATAL ANY
     )
diff --git a/shedskin/cmake.py b/shedskin/cmake.py
index cf69f158..fccb8b8a 100644
--- a/shedskin/cmake.py
+++ b/shedskin/cmake.py
@@ -617,6 +617,7 @@ class CMakeBuilder:
         else:
             self.source_dir = pathlib.Path.cwd().parent
             self.build_dir = self.source_dir / "build"
+        self.build_dir = self.build_dir.resolve()
         self.tests = sorted(glob.glob("./test_*/test_*.py", recursive=True))
         self.log = logging.getLogger(self.__class__.__name__)
 
@@ -678,7 +679,8 @@ class CMakeBuilder:
     def cmake_config(self, options: list[str], generator: Optional[str] = None) -> None:
         """CMake configuration phase"""
         opts = " ".join(options)
-        cfg_cmd = f"cmake {opts} -S {self.source_dir} -B {self.build_dir}"
+        assert isinstance(self.build_dir, pathlib.Path)
+        cfg_cmd = f"cmake {opts} -S {self.source_dir} -B {self.build_dir.resolve()}"
         if generator:
             cfg_cmd += ' -G "{generator}"'
         self.log.info(cfg_cmd)
@@ -875,3 +877,4 @@ class TestRunner(CMakeBuilder):
         self.build_dir = pathlib.Path("build")
         self.tests = sorted(glob.glob("./test_*/test_*.py", recursive=True))
         self.log = logging.getLogger(self.__class__.__name__)
+
diff --git a/shedskin/subprocess_utils.py b/shedskin/subprocess_utils.py
index 804e657d..6d60cb0c 100644
--- a/shedskin/subprocess_utils.py
+++ b/shedskin/subprocess_utils.py
@@ -59,6 +59,7 @@ def run_command(
         Only use shell=True when you need shell features like pipes or
         environment variable expansion, and ensure cmd is trusted.
     """
+    print(f"cmd: {cmd}")
     try:
         result = subprocess.run(
             cmd,
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 92dee838..ba0e0e2c 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.18.4)
 
-project(shedskin_products LANGUAGES CXX)
+project(test_ops_assignment_project LANGUAGES CXX)
 
 set(CMAKE_CXX_STANDARD 17)
 
@@ -9,14 +9,34 @@ set(CMAKE_CXX_STANDARD 17)
 
 option(DEBUG "Enable debug output" OFF)
 
+option(SIMPLE_PROJECT "Project has a single target and no subprojects" OFF)
+
 option(BUILD_EXECUTABLE "Build executable" ON)
 option(BUILD_EXTENSION "Build python extension" OFF)
 option(BUILD_TEST "Build test" ON)
-
+option(DISABLE_EXECUTABLES "Disable building executables" OFF)
+option(DISABLE_EXTENSIONS "Disable building extensions" OFF)
+option(COLLECT_STATS "Collect stats" OFF)
 option(ENABLE_SPM "Enable shedskin dependency management" OFF)
 option(ENABLE_CONAN "Enable conan for dependency management" OFF)
 option(ENABLE_EXTERNAL_PROJECT "Enable ExternalProject for dependency mgmt" OFF)
 
+option(ENABLE_WARNINGS "Enable -Wall type of warnings" ON)
+
+# -----------------------------------------------------------------------------
+# reporting
+
+message("Project: " ${PROJECT_NAME})
+message("Build type: " ${CMAKE_BUILD_TYPE})
+message("")
+
+# -----------------------------------------------------------------------------
+# warnings
+
+if("${PROJECT_SOURCE_DIR}" STREQUAL "${PROJECT_BINARY_DIR}")
+   message(FATAL_ERROR "In-source builds are not allowed.")
+endif()
+
 # -----------------------------------------------------------------------------
 # find python executable and shedskin package path
 
@@ -48,7 +68,12 @@ include_directories(
 # dependency mgmt solutions
 
 if(ENABLE_SPM)
-    set(SPM_DEPS ${CMAKE_BINARY_DIR}/deps)
+    execute_process(
+        COMMAND ${Python_EXECUTABLE} -c "from shedskin import cmake; cmake.user_cache_dir()"
+        OUTPUT_VARIABLE SPM_DEPS
+        COMMAND_ERROR_IS_FATAL ANY
+    )
+    message(STATUS "SPM_DEPS: ${SPM_DEPS}")
     set(SPM_LIB_DIRS ${SPM_DEPS}/lib)
     set(SPM_INCLUDE_DIRS ${SPM_DEPS}/include)
 
@@ -87,7 +112,7 @@ endif()
 # includes
 list(APPEND CMAKE_MODULE_PATH "${SHEDSKIN}/resources/cmake")
 include(utils)
-include(fn_add_shedskin_product)
+include(fn_add_shedskin_product)    # can build both executable and extension
 
 # -----------------------------------------------------------------------
 # any 'test_*' folders are automatically treated as tests
@@ -118,3 +143,4 @@ endif()
 
 # -----------------------------------------------------------------------
 # add subdirectories here
+add_subdirectory(test_ops_assignment)
diff --git a/tests/test_ops_assignment/CMakeLists.txt b/tests/test_ops_assignment/CMakeLists.txt
index 1b384ec7..55931daa 100644
--- a/tests/test_ops_assignment/CMakeLists.txt
+++ b/tests/test_ops_assignment/CMakeLists.txt
@@ -1,4 +1,6 @@
 add_shedskin_product(
+    BUILD_EXECUTABLE
+    MAIN_MODULE test_ops_assignment.py
     SYS_MODULES
         re
-)
+)
\ No newline at end of file
